controllers/clustergroupupgrade_controller.go

```diff
@@ -281,24 +266,189 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
 	return nextReconcile, nil
 }
 
+func (r *ClusterGroupUpgradeReconciler) initializeRemediationPolicyForBatch(
+	clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) {
+
+	if clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.IsZero() {
+		clusterGroupUpgrade.Status.Status.RemediationPlanForBatch = make(map[string]int)
+		batchIndex := clusterGroupUpgrade.Status.Status.CurrentBatch - 1
+
+		// By default, assume all clusters start applying all policies from the first policy.
+		for _, batchClusterName := range clusterGroupUpgrade.Status.RemediationPlan[batchIndex] {
+			clusterGroupUpgrade.Status.Status.RemediationPlanForBatch[batchClusterName] = 0
+		}
+
+		// Set the time for when the batch started updating.
+		clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt = metav1.Now()
+
+		r.Log.Info(">>>> InitializeRemediationPolicyForBatch ", "remediationPlanForBatch", clusterGroupUpgrade.Status.Status.RemediationPlanForBatch)
+		r.Log.Info("NEW CurrentBatchStartedAt", "clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt", clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt)
+	}
+}
+
+func (r *ClusterGroupUpgradeReconciler) getNextRemediationPoliciesForBatch(
+	ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) (error, bool) {
+	r.Log.Info("\n === In getRemediationPlanForBatch ===", "batchIndex", clusterGroupUpgrade.Status.Status.CurrentBatch)
+	r.Log.Info("CURRENT RemediationPlanForBatch", "plan", clusterGroupUpgrade.Status.Status.RemediationPlanForBatch)
+
+	var isBatchComplete bool = false
+	batchIndex := clusterGroupUpgrade.Status.Status.CurrentBatch - 1
+
+	for _, batchClusterName := range clusterGroupUpgrade.Status.RemediationPlan[batchIndex] {
+		// Get the index and name of the current policy being applied for the current cluster in the batch.
+		currentPolicyIndex := clusterGroupUpgrade.Status.Status.RemediationPlanForBatch[batchClusterName]
+		currentManagedPolicyName := clusterGroupUpgrade.Spec.ManagedPolicies[currentPolicyIndex]
+		r.Log.Info("Current policy name for cluster ", "cluster name", batchClusterName, "policyName", currentManagedPolicyName)
+
+		err, currentManagedPolicy := r.getPolicyByName(ctx, currentManagedPolicyName, clusterGroupUpgrade.Namespace)
+		if err != nil {
+			// Maybe continue to next cluster in batch or just move to the next policy???
```

Comment:
In here it would only be an error with getting a policy, that's why I was a bit unsure. There are many other places where we return errors in case we cannot get a managedCluster, a placement rule/binding. I think it would be worth having a better handling of these errors, propagating the main error to the main reconcile loop and include it in the status somehow.
I propose we add this to the cleanup as well.